File: python_pure/src/volarb_py/implied_correlation.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Final

MINIMUM_CONSTITUENTS: Final[int] = 2
MINIMUM_WEIGHT_TOTAL: Final[float] = 1e-12
MINIMUM_OFF_DIAGONAL: Final[float] = 1e-18
PERFECT_CORRELATION: Final[float] = 1.0
WEIGHT_TOTAL_TOLERANCE: Final[float] = 1e-9


class InvalidBasketError(ValueError):
    pass


@dataclass(frozen=True)
class BasketConstituent:
    symbol: str
    weight: float
    volatility: float


@dataclass(frozen=True)
class BasketMoments:
    weighted_volatility: float
    diagonal_variance: float
    off_diagonal_variance: float
    concentration: float


@dataclass(frozen=True)
class CorrelationReport:
    constituent_count: int
    index_volatility: float
    weighted_volatility: float
    concentration: float
    clean_correlation: float
    diagonal_bias: float
    dirty_correlation: float
    lowest_admissible_correlation: float
    is_admissible: bool
    exceeds_perfect_correlation: bool


def validate_basket(constituents: list[BasketConstituent]) -> None:
    if len(constituents) < MINIMUM_CONSTITUENTS:
        raise InvalidBasketError(
            f"a basket needs at least {MINIMUM_CONSTITUENTS} constituents, got {len(constituents)}"
        )
    total = 0.0
    for constituent in constituents:
        if constituent.weight <= 0.0:
            raise InvalidBasketError(
                f"{constituent.symbol} carries a weight of {constituent.weight}, which is not positive"
            )
        if constituent.volatility <= 0.0:
            raise InvalidBasketError(
                f"{constituent.symbol} carries a volatility of {constituent.volatility}, "
                f"which is not positive"
            )
        total += constituent.weight
    if abs(total - 1.0) > WEIGHT_TOTAL_TOLERANCE:
        raise InvalidBasketError(f"the constituent weights sum to {total} rather than one")


def moments_of(constituents: list[BasketConstituent]) -> BasketMoments:
    weighted = 0.0
    diagonal = 0.0
    for constituent in constituents:
        contribution = constituent.weight * constituent.volatility
        weighted += contribution
        diagonal += contribution * contribution
    off_diagonal = weighted * weighted - diagonal
    return BasketMoments(
        weighted_volatility=weighted,
        diagonal_variance=diagonal,
        off_diagonal_variance=off_diagonal,
        concentration=diagonal / (weighted * weighted),
    )


def lowest_admissible_correlation(constituent_count: int) -> float:
    return -1.0 / (constituent_count - 1)


def basket_volatility(constituents: list[BasketConstituent], correlation: float) -> float:
    validate_basket(constituents)
    floor = lowest_admissible_correlation(len(constituents))
    if correlation < floor or correlation > PERFECT_CORRELATION:
        raise InvalidBasketError(
            f"a correlation of {correlation} is outside the admissible range "
            f"[{floor}, {PERFECT_CORRELATION}] for {len(constituents)} constituents"
        )
    moments = moments_of(constituents)
    variance = moments.diagonal_variance + correlation * moments.off_diagonal_variance
    return math.sqrt(max(variance, 0.0))


def imply_correlation(constituents: list[BasketConstituent], index_volatility: float) -> CorrelationReport:
    validate_basket(constituents)
    if index_volatility <= 0.0:
        raise InvalidBasketError(f"the index volatility must be positive, got {index_volatility}")
    moments = moments_of(constituents)
    if moments.off_diagonal_variance < MINIMUM_OFF_DIAGONAL:
        raise InvalidBasketError(
            "the off diagonal variance is nothing, so no correlation can be implied from it"
        )
    index_variance = index_volatility * index_volatility
    clean = (index_variance - moments.diagonal_variance) / moments.off_diagonal_variance
    bias = moments.concentration * (PERFECT_CORRELATION - clean)
    floor = lowest_admissible_correlation(len(constituents))
    return CorrelationReport(
        constituent_count=len(constituents),
        index_volatility=index_volatility,
        weighted_volatility=moments.weighted_volatility,
        concentration=moments.concentration,
        clean_correlation=clean,
        diagonal_bias=bias,
        dirty_correlation=clean + bias,
        lowest_admissible_correlation=floor,
        is_admissible=floor <= clean <= PERFECT_CORRELATION,
        exceeds_perfect_correlation=clean > PERFECT_CORRELATION,
    )


def index_sensitivities(constituents: list[BasketConstituent], correlation: float) -> list[float]:
    index_volatility = basket_volatility(constituents, correlation)
    moments = moments_of(constituents)
    sensitivities: list[float] = []
    for constituent in constituents:
        contribution = constituent.weight * constituent.volatility
        others = moments.weighted_volatility - contribution
        sensitivities.append(constituent.weight * (contribution + correlation * others) / index_volatility)
    return sensitivities


def dispersion_vega_weights(
    constituents: list[BasketConstituent], correlation: float, index_vega: float
) -> list[float]:
    return [index_vega * sensitivity for sensitivity in index_sensitivities(constituents, correlation)]

File: cpp_core/src/implied_correlation.cpp
#include "volarb/implied_correlation.hpp"

#include <cmath>

namespace volarb {

void validate_basket(const std::vector<BasketConstituent>& constituents) {
    if (constituents.size() < minimum_constituents) {
        throw InvalidBasketError("a basket needs at least " + std::to_string(minimum_constituents) +
                                 " constituents, got " + std::to_string(constituents.size()));
    }
    double total = 0.0;
    for (const BasketConstituent& constituent : constituents) {
        if (constituent.weight <= 0.0) {
            throw InvalidBasketError(constituent.symbol + " carries a weight of " +
                                     std::to_string(constituent.weight) + ", which is not positive");
        }
        if (constituent.volatility <= 0.0) {
            throw InvalidBasketError(constituent.symbol + " carries a volatility of " +
                                     std::to_string(constituent.volatility) + ", which is not positive");
        }
        total += constituent.weight;
    }
    if (std::abs(total - 1.0) > weight_total_tolerance) {
        throw InvalidBasketError("the constituent weights sum to " + std::to_string(total) +
                                 " rather than one");
    }
}

BasketMoments moments_of(const std::vector<BasketConstituent>& constituents) {
    double weighted = 0.0;
    double diagonal = 0.0;
    for (const BasketConstituent& constituent : constituents) {
        const double contribution = constituent.weight * constituent.volatility;
        weighted += contribution;
        diagonal += contribution * contribution;
    }
    const double off_diagonal = weighted * weighted - diagonal;
    return BasketMoments{weighted, diagonal, off_diagonal, diagonal / (weighted * weighted)};
}

double lowest_admissible_correlation(std::size_t constituent_count) {
    return -1.0 / static_cast<double>(constituent_count - 1);
}

double basket_volatility(const std::vector<BasketConstituent>& constituents, double correlation) {
    validate_basket(constituents);
    const double floor = lowest_admissible_correlation(constituents.size());
    if (correlation < floor || correlation > perfect_correlation) {
        throw InvalidBasketError("a correlation of " + std::to_string(correlation) +
                                 " is outside the admissible range [" + std::to_string(floor) + ", " +
                                 std::to_string(perfect_correlation) + "] for " +
                                 std::to_string(constituents.size()) + " constituents");
    }
    const BasketMoments moments = moments_of(constituents);
    const double variance = moments.diagonal_variance + correlation * moments.off_diagonal_variance;
    return std::sqrt(variance > 0.0 ? variance : 0.0);
}

CorrelationReport imply_correlation(const std::vector<BasketConstituent>& constituents,
                                    double index_volatility) {
    validate_basket(constituents);
    if (index_volatility <= 0.0) {
        throw InvalidBasketError("the index volatility must be positive, got " +
                                 std::to_string(index_volatility));
    }
    const BasketMoments moments = moments_of(constituents);
    if (moments.off_diagonal_variance < minimum_off_diagonal_variance) {
        throw InvalidBasketError(
            "the off diagonal variance is nothing, so no correlation can be implied from it");
    }
    const double index_variance = index_volatility * index_volatility;
    const double clean = (index_variance - moments.diagonal_variance) / moments.off_diagonal_variance;
    const double bias = moments.concentration * (perfect_correlation - clean);
    const double floor = lowest_admissible_correlation(constituents.size());
    return CorrelationReport{static_cast<int>(constituents.size()),
                             index_volatility,
                             moments.weighted_volatility,
                             moments.concentration,
                             clean,
                             bias,
                             clean + bias,
                             floor,
                             floor <= clean && clean <= perfect_correlation,
                             clean > perfect_correlation};
}

std::vector<double> index_sensitivities(const std::vector<BasketConstituent>& constituents,
                                        double correlation) {
    const double index_volatility = basket_volatility(constituents, correlation);
    const BasketMoments moments = moments_of(constituents);
    std::vector<double> sensitivities;
    sensitivities.reserve(constituents.size());
    for (const BasketConstituent& constituent : constituents) {
        const double contribution = constituent.weight * constituent.volatility;
        const double others = moments.weighted_volatility - contribution;
        sensitivities.push_back(constituent.weight * (contribution + correlation * others) /
                                index_volatility);
    }
    return sensitivities;
}

std::vector<double> dispersion_vega_weights(const std::vector<BasketConstituent>& constituents,
                                            double correlation, double index_vega) {
    std::vector<double> weights = index_sensitivities(constituents, correlation);
    for (double& weight : weights) {
        weight *= index_vega;
    }
    return weights;
}

}
